File: e2e/java/src/test/java/dev/kreuzberg/e2e/ContentTest.java
package dev.kreuzberg.e2e;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import dev.kreuzberg.kreuzcrawl.Kreuzcrawl;

/** E2e tests for category: content. */
class ContentTest {
    @Test
    void testContent204NoContent() throws Exception {
        // Handles 204 No Content response gracefully
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(204, result.statusCode());
        assertTrue(result.html().isEmpty(), "expected empty value");
    }

    @Test
    void testContentCharsetIso8859() throws Exception {
        // Handles ISO-8859-1 encoded page correctly
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals("iso-8859-1", result.detectedCharset().orElse(""));
    }

    @Test
    void testContentEmptyBody() throws Exception {
        // Handles 200 response with empty body gracefully
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(200, result.statusCode());
    }

    @Test
    void testContentGzipCompressed() throws Exception {
        // Handles response with Accept-Encoding gzip negotiation
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertFalse(result.html().isEmpty(), "expected non-empty value");
        assertEquals(200, result.statusCode());
    }

    @Test
    void testContentLargePageLimit() throws Exception {
        // Respects max body size limit and truncates or skips oversized pages
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertTrue(result.bodySize() < 1025, "expected < 1025");
    }

    @Test
    void testContentMainOnly() throws Exception {
        // Extracts only main content area, excluding nav, sidebar, footer
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.mainContentOnly());
    }

    @Test
    void testContentPdfNoExtension() throws Exception {
        // Detects PDF content by Content-Type header when URL has no .pdf extension
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals(true, result.isPdf());
    }

    @Test
    void testContentRemoveTags() throws Exception {
        // Removes specified HTML elements by CSS selector before processing
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertFalse(result.html().isEmpty(), "expected non-empty value");
    }

    @Test
    void testContentUtf8Bom() throws Exception {
        // Handles UTF-8 content with BOM marker correctly
        var engine = Kreuzcrawl.createEngine(null);
        var result = Kreuzcrawl.scrape(engine, "");
        assertEquals("utf-8", result.detectedCharset().orElse(""));
        assertFalse(result.html().isEmpty(), "expected non-empty value");
    }

}

File: e2e/go/content_test.go
// E2e tests for category: content
package e2e_test

import (
	"strings"
	"testing"

	pkg "github.com/kreuzberg-dev/kreuzcrawl"
)

func Test_Content204NoContent(t *testing.T) {
	// Handles 204 No Content response gracefully
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.StatusCode != 204 {
		t.Errorf("equals mismatch: got %q", result.StatusCode)
	}
	if len(result.Html) != 0 {
		t.Errorf("expected empty value, got %q", result.Html)
	}
}

func Test_ContentCharsetIso8859(t *testing.T) {
	// Handles ISO-8859-1 encoded page correctly
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'content.detected_charset' not available on result type
}

func Test_ContentEmptyBody(t *testing.T) {
	// Handles 200 response with empty body gracefully
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %q", result.StatusCode)
	}
}

func Test_ContentGzipCompressed(t *testing.T) {
	// Handles response with Accept-Encoding gzip negotiation
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Html) == 0 {
		t.Errorf("expected non-empty value")
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %q", result.StatusCode)
	}
}

func Test_ContentLargePageLimit(t *testing.T) {
	// Respects max body size limit and truncates or skips oversized pages
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'content.body_size' not available on result type
}

func Test_ContentMainOnly(t *testing.T) {
	// Extracts only main content area, excluding nav, sidebar, footer
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'content.main_content_only' not available on result type
}

func Test_ContentPdfNoExtension(t *testing.T) {
	// Detects PDF content by Content-Type header when URL has no .pdf extension
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	_, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'content.is_pdf' not available on result type
}

func Test_ContentRemoveTags(t *testing.T) {
	// Removes specified HTML elements by CSS selector before processing
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Html) == 0 {
		t.Errorf("expected non-empty value")
	}
}

func Test_ContentUtf8Bom(t *testing.T) {
	// Handles UTF-8 content with BOM marker correctly
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	// skipped: field 'content.detected_charset' not available on result type
	if len(result.Html) == 0 {
		t.Errorf("expected non-empty value")
	}
}
